### PR TITLE
fix(lib/blocktree): removes the inconsistency to choose a deepest leaf

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
### Quick explanation

Given the following block tree representation

```
           /-> d -> f
a -> b -> c -> d
      \-> c -> d -> f
```

considering the upper `f` and lower `f` have the same number and arrival time, the function `blocktree.leaves. deepestLeaf()` can return inconsistent output since `sync.Map` doesn't guarantee the same sequence every time.

So this inconsistency causes the problem in the test `TestBlockTree_GetHashByNumber` for the following reasons:
1. If the function `blocktree.GetHashByNumber` needs to find the block with number `c` and takes the upper `f` as its deepest leaf then the output will be the middle `c` block
2. However, the function `blocktree.IsDescendantOf` receives the lower `f` and middle `c` as parameters to check if `f` is descendant of `c` which is false because lower `f` is descendant of **lower `c`**

### Updates
- This PR adds a field called `currentDeepestLeaf` at `leafMap` struct which holds the deepest node and ignores the others one whose have the same number and arrival time

## Tests

```
WIP
```

## Issues

- Related #1975

## Primary Reviewer

- @qdm12

### CODE REVIEW

File: lib/blocktree/blocktree_test.go [hunk 1]
```diff
@@ -474,6 +477,111 @@ func TestBlockTree_GetHashByNumber(t *testing.T) {
...
+	const fixedArrivalTime = 99
+	const deep = 8
+
+	// create a base tree with a fixed amount of blocks
+	// and all block with the same arrival time
+
+	/**
+	base tree and nodes representation, all with the same arraival time and all
+	the leaves has the same number (8) the numbers in the right represents the order
+	the nodes are inserted into the blocktree.
```
Comment (1garo) [inline lib/blocktree/blocktree_test.go:495]: typo at arraival

Comment (1garo) [inline lib/blocktree/blocktree_test.go:495]: typo at arraival

File: lib/blocktree/blocktree_test.go [hunk 2]
```diff
@@ -474,6 +477,111 @@ func TestBlockTree_GetHashByNumber(t *testing.T) {
...
+		// the last block on the base tree should not generates a branch
+		if i < deep {
+			branches = append(branches, testBranch{
+				hash:   hash,
+				number: bt.getNode(hash).number,
+			})
+		}
+	}
+
+	// create all the branch nodes with the same arraival time
```
Comment (1garo) [inline lib/blocktree/blocktree_test.go:529]: typo here too

**Review by noot — APPROVED** [review]
this seems like it should work, thanks for the fix!

**Review by edwardmack — APPROVED** [review]
Great work, lgtm.

**Pull request merged**